[wpilibj] Add BooleanMajorityFilter

Starlight220 opened the pull request:
A MedianFilter for boolean values, with tests. The filter also has a ctor overload for changing the majority percentage. Resolves #2482.

Starlight220 commented:
What is the correct import order? I truly seem to not be able to find the correct one...

AustinShalit commented:
The rule that describes import order is here:

https://github.com/wpilibsuite/allwpilib/blob/e5935a4737e5a6f92aeae5d43c7b7a7f26a191ee/styleguide/checkstyle.xml#L88-L96

There are 7 standard groups that are separated by new lines.

Starlight220 commented:
Can someone please run the testbench?

AustinShalit commented:
@Starlight220 We were thinking about this PR and were hoping you could help us find a specific example of how this would be used. Did you have any ideas when composing this PR?

Starlight220 commented:
The idea is from a mechanism my team used this year. We had a flywheel, and we wanted to bring a ball from the indexer only if the flywheel was above a certain RPM. As this was a relatively inconsistent boolean reading, we developed a mechanism to get a more consistent reading. We implemented it [here](https://github.com/ExcaliburFRC/InfiniteRecharge2020/blob/master/src/main/java/frc/robot/subsystems/BooleanAverager.java)
and it is used mostly [here](https://github.com/ExcaliburFRC/InfiniteRecharge2020/blob/master/src/main/java/frc/robot/subsystems/Shooter.java) (there isn't documentation, it really is something my team needs to work on).

For this PR, I adjusted the implementation a bit, both to patch a few problems I saw with our implementation and for consistency with the other Filter types already in WPILib.

calcmogul commented:
I just looked over your team's code, and I think there's a more appropriate solution to the problem than BooleanMajorityFilter based on what the root cause of your performance issue may be. From prior experience, it could be noisy measurements, inaccurate feedforward causing steady-state error, or poor controller tuning.

First, I noticed you have no filtering on the shooter velocity, and velocity measurements tend to be very noisy. You can filter the measurements with a single-pole IIR filter (see LinearFilter in WPILib).

Second, you're using an integrator to account for steady-state error. This tells me your feedforward isn't accurate enough. You may need to retune it with frc-characterization. You should only need a velocity feedforward and proportional feedback for a flywheel. (By the way, WPILib's PIDController supports integrator clamping with setIntegratorRange(). I assume lack of that feature is why you rolled your own PID controller. `edu.wpi.first.wpiutil.math` also has MathUtil.clmap().)

Third, flywheels can have aggressive proportional gains if you can eliminate measurement time delay (this is a good thing for shot recovery). Since you're grabbing velocity from a Talon SRX, the velocity data is being sent in a CAN status frame every 10ms by default. Depending on your flywheel's dynamics, that might be too slow. That is, you have to use lower proportional gains with poor disturbance rejection because the higher gains cause oscillation. For example, 3512's flywheel had a shot recovery time of 20ms due to the gearing we used with a control action computed every 5ms. 10ms of delay wasn't workable for us, and we could decrease the status frame period without causing dropped CAN frames, so we had to plug an encoder back into the roboRIO DIO. Hopefully, the root cause of your issue can be fixed with something less drastic than this.

Plots of the setpoints, measurements, and control inputs over time are very helpful when diagnosing issues like this.

Starlight220 commented:
Although I agree with @calcmogul that we may have designed our flywheel control poorly, the use case of minimizing limitswitch or button inconsistencies remains. I am aware that the Debouncer class exists and works by time, I am proposing a count-based option.

Starlight220 commented:
On a different note, @calcmogul : 

> (By the way, WPILib's PIDController supports integrator clamping with setIntegratorRange(). I assume lack of that feature is why you rolled your own PID controller.

We had a pretty bad experience with the old WPILib PID setup, and I guess it rolled over to the new one. Maybe we'll try it next year!

> `edu.wpi.first.wpiutil.math` also has MathUtil.clamp().)

We dislike the fact that both min and max values need to be specified, and for most of our uses we prefer having `clamp(val, bound)` that theoretically translates into `clamp(val, min = -bound, max = bound)`. I suggested adding such an overload in #2412, but it got rejected.

calcmogul commented:
> ... the use case of minimizing limitswitch or button inconsistencies remains.

Use DigitalGlitchFilter for debouncing limit switches. Have you needed debouncing on joystick buttons? I've never needed it for Logitech joysticks. Joystick buttons only being sent every 20ms might be a factor.

> I am aware that the Debouncer class exists and is time-based, I am proposing a count-based option.

Well, we shouldn't have both. The Debouncer class should support whichever method we think makes the most sense for our users, if one exists. I think specifying "the amount of time the value has to stay constant" is more intuitive than working out how many samples that is. It's also more robust to varying sample rates.
